Brokerages/Alpaca/AlpacaBrokerage.cs

```diff
@@ -0,0 +1,415 @@
+        {
+            CheckRateLimiting();
+
+            var task = _restClient.GetAccountAsync();
+            var balance = task.SynchronouslyAwaitTaskResult();
+
+            return new List<Cash>
+            {
+                new Cash("USD",
+                    balance.TradableCash,
+                    1m)
+            };
+        }
+
+        /// <summary>
+        /// Gets all holdings for the account
+        /// </summary>
+        /// <returns>The current holdings from the account</returns>
+        public override List<Holding> GetAccountHoldings()
+        {
+            CheckRateLimiting();
+
+            var task = _restClient.ListPositionsAsync();
+            var holdings = task.SynchronouslyAwaitTaskResult();
+
+            var qcHoldings = holdings.Select(ConvertHolding).ToList();
+
+            // Set MarketPrice in each Holding
+            var alpacaSymbols = qcHoldings
+                .Select(x => x.Symbol.Value)
+                .ToList();
+
+            if (alpacaSymbols.Count > 0)
+            {
+                var quotes = GetRates(alpacaSymbols);
+                foreach (var holding in qcHoldings)
+                {
+                    var alpacaSymbol = holding.Symbol;
+                    Tick tick;
+                    if (quotes.TryGetValue(alpacaSymbol.Value, out tick))
+                    {
+                        holding.MarketPrice = (tick.BidPrice + tick.AskPrice) / 2;
+                    }
+                }
+            }
+
+            return qcHoldings;
+        }
+
+        /// <summary>
+        /// Gets all open orders on the account.
+        /// NOTE: The order objects returned do not have QC order IDs.
+        /// </summary>
+        /// <returns>The open orders returned from Alpaca</returns>
+        public override List<Order> GetOpenOrders()
+        {
+            CheckRateLimiting();
+
+            var task = _restClient.ListOrdersAsync();
+            var orders = task.SynchronouslyAwaitTaskResult();
+
+            return orders.Select(ConvertOrder).ToList();
+        }
+
+        /// <summary>
+        /// Places a new order and assigns a new broker ID to the order
+        /// </summary>
+        /// <param name="order">The order to be placed</param>
+        /// <returns>True if the request for a new order has been placed, false otherwise</returns>
+        public override bool PlaceOrder(Order order)
+        {
+            const int orderFee = 0;
+            order.PriceCurrency = "USD";
+
+            lock (_locker)
+            {
+                try
+                {
+                    var apOrder = GenerateAndPlaceOrder(order);
+                    order.BrokerId.Add(apOrder.OrderId.ToString());
+                }
+                catch (Exception e)
+                {
+                    var errorMessage = $"Error placing order: {e.Message}";
+
+                    OnOrderEvent(
+                        new OrderEvent(order, DateTime.UtcNow, 0, "Alpaca Order Event")
```

Comment:
Thinking maybe the `lock` statement should be inside the `try` `catch` to avoid looking over the user code, called through the OrderEvent callback.